### ISSUE TITLE
transaction feature: multiple *dependent* mutation operations within a single mutation request

### ISSUE THREAD

**Issue by hartig** [LiUGraphQL/woo.sh#52]
While the generated servers can already support mutation requests that contain multiple operations that are independent from one another, the purpose of this issue is to also support mutation requests that contain multiple operations that depend on one another. The typical use case for such a functionality is the following: A client may want to create two objects within a single transaction where some value to be used for creating the second object is a value that is auto-generated when creating the first object (e.g., an ID). In terms of the GraphQL request language, this may look like the following:
```graphql
mutation {
  createXYZ( data: {
        name: "D"
     }
  )
  {
    id @export(as: "idOfD")
    name
  }

  createXYZ( data: {
      name: "C",
      neighbors: [
        { connect: $idOfD }
      ]
    }
  )
  {
    id
    name
  }
}
```
Notice that this mutation request contains two separate create operations where the second one uses the ID created by the first one.

Notice also that this example uses a special directive called `@export`, which is necessary to establish some form of variables (like the variable `$idOfD` in this example). While there is nothing for this in the GraphQL spec, there is a [discussion of a related feature for the next version of the spec](https://github.com/graphql/graphql-spec/issues/377).

**Comment by keski** [comment]
Note that commit above was simply tagged incorrectly and has nothing to do with this issue.

**Comment by keski** [comment]
Based on the discussion regarding dynamic variables in the link above, I'd suggest we start looking in to this for some simple cases and only consider it for mutations for now. I also think we need to stay within the existing spec and not attempt any extension to the GraphQL syntax.

I also think implementing this is relatively easy to do our implementation, since we already execute mutations sequentially. 

A full example, containing both the schema and a mutation, that can be used as a starting point for further discussion is provided below.
```graphql
directive @export(as: String!) on FIELD

type Blogger {
  name: String!
}

type Blog {
  text: String!
  author: Blogger
}
```
*Create a blogger and then create a blog with the blogger as the author.*
```graphql
mutation ($bloggerId:ID) {
  createBlogger(data:{
    name: "John Doe"
  }){
    id @export(as: "bloggerId")
  }
  createBlog(data: {
    text:"Some text"
    author: {
      connect: $bloggerId
    }
  }){
    id
    author {
      id
    }
  }
}
```

**Comment by keski** [comment]
https://github.com/LiUGraphQL/woo.sh/tree/52-transaction-feature-dependent-ops

**Comment by hartig** [comment]
In our call we decided to restrict the use of this `@export` directive to simplify our lives. The restriction is that `@export` can be used only for the scalar-typed fields of the immediate return type of the mutation operation. As an example, consider the following mutation request.
```graphql
mutation {
  createBlog(
    data: {
      text: "Some text"
      author: {
        connect: "8713"
      }
     }
  ) {
    id       @export(as: "bloggerID")     # this is allowed
    author   @export(as: "authorObject")  # this is *not* allowed
    {
      id     @export(as: "authorID")  # this is *not* allowed
    }
  }
}
```

**Comment by keski** [comment]
In order to make this work, we need to disable the validation of unused variables [specifiedRules.js](https://github.com/graphql/graphql-js/blob/master/src/validation/specifiedRules.js) in grahpql-js. Since we are already making a modification to the graphql lib to make transactions work, I'll create a fork of graphql-js as previously discussed.

**Comment by keski** [comment]
Following the latest updates of [LiUGraphQL/graphql-js](https://github.com/LiUGraphQL/graphql-js/tree/woosh) the library has been updated to skip some validation steps for exported variables. The directives are also added to the exported variable definitions, which allows them to be accessed by the driver. The ArangoDB driver has been updated to support the export directive for creating objects, updating objects, and creating edges. Type validation of the exported field value remains to be implemented.

**Comment by keski** [comment]
The validation of exported variables became quite tangled up with driver code, making the code much harder to maintain. Validation has therefore been integrated as a separate rule in the GraphQL validation pipeline. This means that the validation is now independent of the particular driver employed, while also providing much more constructive user feedback.

The relevant graphql-js files include: [ExportVariablesRule.js](https://github.com/LiUGraphQL/graphql-js/blob/woosh/src/validation/rules/ExportVariablesRule.js), [specifiedRules.js](https://github.com/LiUGraphQL/graphql-js/blob/woosh/src/validation/specifiedRules.js), and [values.js](https://github.com/LiUGraphQL/graphql-js/blob/woosh/src/execution/values.js). The limitation of not allowing export of nested fields is currently treated as a driver limitation (rather than a strict limitation) and these errors are therefore raised from the ArangoDB driver.

**Comment by keski** [comment]
Resolved (see #99)

**Comment by hartig** [comment]
Wohoo!! It's great that this feature is implemented. Thanks for your effort!!
